group/assignment-6/src/cars/RearrangingCars.java

```diff
@@ -0,0 +1,340 @@
+            "The start state of the parking lot violates the constraints.");
+      }
+    }
+  }
+
+  // to be potentially modified based on allowedParking constraints format (to be decided)
+  private static void validateAllowedParkingConstraints(
+      ParkingLot parkingLot, Map<ParkingSpot, Set<Car>> allowedParking) throws RuntimeException {
+    if (allowedParking == null) {
+      throw new IllegalArgumentException("You must enter defined parking constraints.");
+    }
+
+    Collection<Car> cars = parkingLot.getCars();
+    Collection<ParkingSpot> spots = parkingLot.getSpots();
+    int numOfSpotsToBeEmpty = 0;
+
+    for (Entry<ParkingSpot, Set<Car>> currentConstraint : allowedParking.entrySet()) {
+      if (currentConstraint.getValue().isEmpty()) {
+        numOfSpotsToBeEmpty++;
+
+        // if we have more than 1 spot that is to be empty, we can't possibly generate a valid
+        // sequence
+        // of moves. either the start state violates the constraints or the final parking
+        // configuration
+        // can't be reached because we only have room for 1 empty spot
+        if (numOfSpotsToBeEmpty == 2) {
+          throw new NoPossibleSequenceException(
+              "The start state of the parking lot violates the constraints.");
+        }
+      }
+
+      if (!spots.contains(currentConstraint.getKey())) {
+        throw new IllegalArgumentException(
+            "Each parking spot in the parking constraints has to be contained in the given parking lot.");
+      }
+
+      if (!cars.containsAll((currentConstraint.getValue()))) {
+        throw new IllegalArgumentException(
+            "All mapped cars in the parking constraints have to be contained in the given parking lot.");
+      }
+    }
+  }
+
+  private static void getAllPossibleSequencesUtil(
+      ParkingLotConfigurations trie,
+      List<List<CarMove>> sequences,
+      Map<Car, ParkingSpot> carsToSpots,
+      List<CarMove> sequence,
+      int numOfCarsInEndPosition,
+      ParkingLot endState) {
+    if (numOfCarsInEndPosition == carsToSpots.size() - 1) {
+      sequences.add(sequence.stream().collect(Collectors.toList()));
+      return;
+    }
+
+    for (Car car : endState.getCars()) {
+      // move current car to the empty spot
+      ParkingSpot originalSpot = carsToSpots.get(car);
+      ParkingSpot emptySpot = carsToSpots.get(null);
+      moveCar(carsToSpots, car);
+      numOfCarsInEndPosition =
+          updateNumOfCarsInEndLocation(numOfCarsInEndPosition, car, carsToSpots, endState);
+      // if this config is new go deeper
+      if (!trie.contains(carsToSpots.values())) {
+        trie.add(carsToSpots.values());
+        CarMove currentCarMove = new CarMove(car, originalSpot, emptySpot);
+        sequence.add(currentCarMove);
+        getAllPossibleSequencesUtil(
+            trie, sequences, carsToSpots, sequence, numOfCarsInEndPosition, endState);
+        sequence.remove(currentCarMove);
```

Comment:
```suggestion
        sequence.remove(sequence.size() - 1);
```

Sequence is a list, so if you remove an object, it will iterate over the entire list to find that object. Here you would always remove the last car move that was added to sequence.

Reply:
oh yeah that bit was like that because it used to be a set. great that you noticed ! will change

Reply:
lol cheers 
yeah I figured that must have been the case:)